container-registry/container-analysis/src/main/java/com/example/containeranalysis/CreateOccurrence.java

```diff
@@ -39,20 +41,29 @@ public static Occurrence createOccurrence(String resourceUrl, String noteId,
 
     Occurrence.Builder occBuilder = Occurrence.newBuilder();
     occBuilder.setNoteName(noteName.toString());
+    occBuilder.setResourceUri(resourceUrl);
     // Associate the Occurrence with the metadata type (should match the parent Note's type)
     // https://cloud.google.com/container-registry/docs/container-analysis#supported_metadata_types
     // Here, we use the type "vulnerability"
-    Details.Builder detailsBuilder = Details.newBuilder();
-    occBuilder.setVulnerability(detailsBuilder);
-    // Attach the occurrence to the associated image uri
-    Resource.Builder resourceBuilder = Resource.newBuilder();
-    resourceBuilder.setUri(resourceUrl);
-    occBuilder.setResource(resourceBuilder);
+    PackageIssue.Builder issueBuilder = PackageIssue.newBuilder();
+    issueBuilder.setAffectedCpeUri("your-uri-here");
+    issueBuilder.setAffectedPackage("your-package-here");
+    Version.Builder affectedVersionBuilder = Version.newBuilder();
+    affectedVersionBuilder.setKind(Version.VersionKind.MINIMUM);
+    issueBuilder.setAffectedVersion(affectedVersionBuilder);
+    Version.Builder fixedVersionBuilder = Version.newBuilder();
+    fixedVersionBuilder.setKind(Version.VersionKind.MAXIMUM);
+    issueBuilder.setFixedVersion(fixedVersionBuilder);
+
+    VulnerabilityOccurrence.Builder vulBuilder = VulnerabilityOccurrence.newBuilder();
+    vulBuilder.addPackageIssue(issueBuilder);
```

Comment:
If anyone has any advice on how to write these complex builder sections in a clean way, that would be much appreciated

Reply:
Chain them together, and use build at the end. 
```
PackageIssue issue = PackageIssue.newBuilder()
                                        .setAffectedCpeUri("your-uri-here")
                                        .setAffectedPackage("your-package-here")
                                        ....
                                        .build()
```

Generally, they replace constructors. See an example from the canonical [here](https://github.com/GoogleCloudPlatform/java-docs-samples/blob/3439c96c7ff8661d28210618615d7d95bc4ad7a8/dlp/src/main/java/dlp/snippets/InspectImageFile.java#L73).

Reply:
thanks, that's much better!